orttraining/orttraining/core/optimizer/megatron_transformer.cc

```diff
@@ -150,44 +171,101 @@ bool MegatronTransformer::PartitionWeightByColumn(const Graph& graph, const Node
+        auto* init_tensor = initial_state_it->second.GetMutable<Tensor>();
+
+        OrtValue partitioned;
+        auto element_type = init_tensor->DataType();
+        TensorShape partition_shape(new_shape);
+        std::unique_ptr<Tensor> p_tensor;
+
+        if (utils::IsPrimitiveDataType<float>(element_type)) {
+          float* data_buffer = init_tensor->MutableData<float>();
+
+          // allocate temporary memory to get the column partitioned state
+          std::vector<float> result_buffer;
+          result_buffer.reserve(element_count);
+          PartitionBufferByColumn(data_buffer, row_count, column_count, column_stride, stride, result_buffer);
+
+          // allocate a new buffer as column partitioning cannot re-use the original
+          // buffer as its non-contiguous read on original buffer
+          auto alloc = execution_provider_.GetAllocator(0, OrtMemTypeDefault);
+          p_tensor = onnxruntime::make_unique<Tensor>(element_type,
+                                                      partition_shape,
+                                                      alloc);
+          float* out_buffer = p_tensor->MutableData<float>();
+          memcpy(out_buffer, result_buffer.data(), sizeof(float) * element_count);
+        } else if (utils::IsPrimitiveDataType<MLFloat16>(element_type)) {
+          MLFloat16* data_buffer = init_tensor->MutableData<MLFloat16>();
+
+          // allocate temporary memory to get the column partitioned state
+          std::vector<MLFloat16> result_buffer;
+          result_buffer.reserve(element_count);
+          PartitionBufferByColumn(data_buffer, row_count, column_count, column_stride, stride, result_buffer);
+
+          // allocate a new buffer as column partitioning cannot re-use the original
+          // buffer as it is a non-contiguous read on original buffer
+          auto alloc = execution_provider_.GetAllocator(0, OrtMemTypeDefault);
+          p_tensor = onnxruntime::make_unique<Tensor>(element_type,
+                                                      partition_shape,
+                                                      alloc);
+          MLFloat16* out_buffer = p_tensor->MutableData<MLFloat16>();
+          memcpy(out_buffer, result_buffer.data(), sizeof(MLFloat16) * element_count);
+        } else {
+          ORT_THROW("Unsupported type: ", element_type, "for initial optimizer moments.");
+        }
+        partitioned.Init(p_tensor.release(),
+                         DataTypeImpl::GetType<Tensor>(),
+                         DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
+        initial_states[moments_prefix] = std::move(partitioned);
+      }
     }
   }
 
-  initializer_partition.set_raw_data(result.data(), element_count * sizeof(float));
-  weight_partition_info_[new_initializer_name].megatron_row_partition = 0;
+  weight_partition_info_[original_name].megatron_row_partition = 0;
+  weight_partition_info_[original_name].view_name = new_initializer_name;
```

Comment:
to be honest I am a bit lost on the naming here. what's the view_name ? view is for zero enabled, right?

Reply:
that is correct, but this field can be used for megatron as well. Changing the name form view_name to partitioned_name to make it clear